src/plugins/stake-plugins/generic/pluginInfo/tarotUtils.ts

```diff
@@ -0,0 +1,317 @@
+      }
+      return parseFloat(formatUnits(bigAmount.amount, bigAmount.decimals))
+    },
+
+    async getValuesFromPrice(
+      changes: typeof this.NO_CHANGES,
+      priceA: number,
+      priceB: number
+    ): Promise<{ valueCollateral: number; valueA: number; valueB: number }> {
+      const [valueCollateralPart, amountAPart, amountBPart] = await Promise.all([
+        this.getDeposited(TarotPoolTokenType.Collateral),
+        this.getBorrowed(TarotPoolTokenType.BorrowableA),
+        this.getBorrowed(TarotPoolTokenType.BorrowableB)
+      ])
+      const valueCollateral = this.parseNumber(valueCollateralPart) + changes.changeCollateral
+      const amountA = this.parseNumber(amountAPart) + changes.changeBorrowedA
+      const amountB = this.parseNumber(amountBPart) + changes.changeBorrowedB
+      const valueA = amountA * priceA
+      const valueB = amountB * priceB
+      return {
+        valueCollateral: valueCollateral > 0 ? valueCollateral : 0,
+        valueA: valueA > 0 ? valueA : 0,
+        valueB: valueB > 0 ? valueB : 0
+      }
+    },
+
+    async getMarketValues(changes: typeof this.NO_CHANGES): Promise<{ valueCollateral: number; valueA: number; valueB: number }> {
+      const [priceA, priceB] = await this.getMarketPriceDenomLP()
+      return await this.getValuesFromPrice(this.NO_CHANGES, priceA, priceB)
+    },
+
+    async getMaxDeleverage(): Promise<number> {
+      if (config.isStable) {
+        const [[reserve0, reserve1], { valueCollateral, valueA, valueB }] = await Promise.all([this.getReserves(), this.getMarketValues(this.NO_CHANGES)])
+        const x2 = reserve0 * reserve0
+        const y2 = reserve1 * reserve1
+        const f = (3 * x2 + y2) / (x2 + 3 * y2)
+        const minRepayA = (valueCollateral * f) / (1 + f) / Math.sqrt(SLIPPAGE_FACTOR)
+        const minRepayB = valueCollateral / (1 + f) / Math.sqrt(SLIPPAGE_FACTOR)
+        if (minRepayA >= valueA && minRepayB >= valueB) {
+          const deposited = await this.getDeposited(TarotPoolTokenType.Collateral)
+          return this.parseNumber(deposited)
+        }
+        if (valueCollateral / Math.sqrt(SLIPPAGE_FACTOR) < valueA + valueB) {
+          return 0
+        }
+        if (minRepayA >= valueA) {
+          return valueA * (1 + 1 / f) * Math.sqrt(SLIPPAGE_FACTOR)
+        } else {
+          return valueB * (1 + f) * Math.sqrt(SLIPPAGE_FACTOR)
+        }
+      }
+      const { valueCollateral, valueA, valueB } = await this.getMarketValues(this.NO_CHANGES)
+      const minRepayPerSide = valueCollateral / 2 / Math.sqrt(SLIPPAGE_FACTOR)
+      if (minRepayPerSide >= valueA && minRepayPerSide >= valueB) {
+        const deposited = await this.getDeposited(TarotPoolTokenType.Collateral)
+        return this.parseNumber(deposited)
+      }
+      if (minRepayPerSide * 2 < valueA + valueB) {
+        return 0
+      }
+      return Math.min(valueA, valueB) * 2 * Math.sqrt(SLIPPAGE_FACTOR)
+    },
+
+    async getDeleverageAmounts(
```

Comment:
Can we give this function a nice, relaxing massage (with comments)?

Reply:
nope (it's copied from tarot.to)